Error Loading Streams

This error and possibly related errors produced this error on the server:
```
APP/0 at Socket.<anonymous> (/home/vcap/app/node_modules/pg/lib/connection.js:117:22)Feb 15, 2017 4:58:11 PM
APP/0 at Socket.Readable.push (_stream_readable.js:111:10)Feb 15, 2017 4:58:11 PM
APP/0/home/vcap/app/node_modules/newrelic/lib/transaction/tracer/index.js:191Feb 15, 2017 4:58:11 PM
APP/0 at emitOne (events.js:77:13)Feb 15, 2017 4:58:11 PM
APP/0 at Socket.wrappedEmit [as emit] (/home/vcap/app/node_modules/newrelic/lib/transaction/tracer/index.js:220:46)Feb 15, 2017 4:58:11 PM
APP/0 at readableAddChunk (_stream_readable.js:153:18)Feb 15, 2017 4:58:11 PM
APP/0error: remaining connection slots are reserved for non-replication superuser connectionsFeb 15, 2017 4:58:11 PM
APP/0 ^Feb 15, 2017 4:58:11 PM
APP/0 at Socket.emit (events.js:169:7)Feb 15, 2017 4:58:11 PM
APP/0 throw errFeb 15, 2017 4:58:11 PM
APP/0 at Socket.wrapped (/home/vcap/app/node_modules/newrelic/lib/transaction/tracer/index.js:183:28)Feb 15, 2017 4:58:11 PM
APP/0 at Connection.parseE (/home/vcap/app/node_modules/pg/lib/connection.js:554:11)Feb 15, 2017 4:58:11 PM
APP/0 at TCP.onread (net.js:537:20)Feb 15, 2017 4:58:11 PM
APP/0npm ERR! Linux 4.4.0-45-genericFeb 15, 2017 4:58:11 PM
APP/0npm ERR! npm v2.15.11Feb 15, 2017 4:58:11 PM
APP/0npm ERR! argv "/home/vcap/app/vendor/node/bin/node" "/home/vcap/app/vendor/node/bin/npm" "start"Feb 15, 2017 4:58:11 PM
APP/0npm ERR! node v4.7.2Feb 15, 2017 4:58:11 PM
APP/0npm ERR! Failed at the backend@0.0.0 start script 'node app.js'.Feb 15, 2017 4:58:11 PM
APP/0npm ERR! Feb 15, 2017 4:58:11 PM
APP/0npm ERR! Tell the author that this fails on your system:Feb 15, 2017 4:58:11 PM
APP/0npm ERR! code ELIFECYCLEFeb 15, 2017 4:58:11 PM
APP/0npm ERR! This is most likely a problem with the backend package,Feb 15, 2017 4:58:11 PM
APP/0npm ERR! not with npm itself.Feb 15, 2017 4:58:11 PM
APP/0npm ERR! backend@0.0.0 start: `node app.js`Feb 15, 2017 4:58:11 PM
APP/0npm ERR! npm bugs backendFeb 15, 2017 4:58:11 PM
APP/0npm ERR! Feb 15, 2017 4:58:11 PM
APP/0npm ERR! You can get information on how to open an issue for this project with:Feb 15, 2017 4:58:11 PM
APP/0npm ERR! There is likely additional logging output above.Feb 15, 2017 4:58:11 PM
APP/0npm ERR! Or if that isn't available, you can get their info via:Feb 15, 2017 4:58:11 PM
APP/0npm ERR! node app.jsFeb 15, 2017 4:58:11 PM
APP/0npm ERR! /home/vcap/app/npm-debug.logFeb 15, 2017 4:58:11 PM
APP/0npm ERR! Please include the following file with any support request:Feb 15, 2017 4:58:11 PM
```

this seems to be a psql error:
`error: remaining connection slots are reserved for non-replication superuser connections`

Which can be resized by increasing `max_connection` size or by figuring out what is wrong with current pooling:
https://stackoverflow.com/questions/11847144/heroku-psql-fatal-remaining-connection-slots-are-reserved-for-non-replication

--- a/backend/config/appVars.js
+++ b/backend/config/appVars.js
@@ -1,4 +1,4 @@
-/* 
+/*
   Copyright 2017 IBM Corporation
 
   Licensed under the Apache License, Version 2.0 (the "License");
@@ -16,16 +16,22 @@
 
 var apn = require("apn");
 var cfenv = require("cfenv");
-var vcapServices = require('./private/VCAP_SERVICES.json')
-var vcapDbService = vcapServices["compose-for-postgresql"] ? vcapServices["compose-for-postgresql"][0] : null
+var vcapServices = require('./private/VCAP_SERVICES.json');
+var vcapDbService = vcapServices["compose-for-postgresql"] ? vcapServices["compose-for-postgresql"][0] : null;
 var appEnv = cfenv.getAppEnv({
     vcap: {
         services: vcapServices
     }
 });
-var dbName =  vcapDbService ? vcapDbService.name : null
+var dbName =  vcapDbService ? vcapDbService.name : null;
 var postgresService = appEnv.getService(dbName);
 var credentials = require('./private/credentials');
+var pg = require('pg');
+var config = parseDBURL(postgresService.credentials.uri);
+// max number of clients in the pool
+config['max'] = 10;
+// how long a client is allowed to remain idle before being closed
+config['idleTimeoutMillis'] = 30000;
 
 var appVars = {
     port: appEnv.port,
@@ -41,13 +47,29 @@ var appVars = {
         clientSecret: credentials.google.clientSecret,
         redirectURL: appEnv.url + "/auth/youtube/callback"
     },
-    postgres: postgresService.credentials,
+    pool: new pg.Pool(config),
     accessTokenKey: credentials.app.accessTokenKey,
     refreshTokenKey: credentials.app.refreshTokenKey,
     apn: new apn.Provider({
         cert: __dirname + '/private/cert.pem',
         key: __dirname + '/private/key.pem'
     })
+};
+
+function parseDBURL(dbURL) {
+    const url = require('url');
+
+    const params = url.parse(dbURL);
+    const auth = params.auth.split(':');
+
+    return {
+        user: auth[0],
+        password: auth[1],
+        host: params.hostname,
+        port: params.port,
+        database: params.pathname.split('/')[1],
+        ssl: true
+    };
 };
 
 module.exports = appVars;

--- a/backend/id/id.service.js
+++ b/backend/id/id.service.js
@@ -15,7 +15,6 @@
 */
 
 var appVars = require('../config/appVars');
-var pg = require('pg');
 var userController = require('../user/user.controller')
 
 var idService = {};

--- a/backend/invites/invites.service.js
+++ b/backend/invites/invites.service.js
@@ -17,7 +17,6 @@
 var apn = require('apn');
 var userController = require('../user/user.controller');
 var appVars = require('../config/appVars');
-var pg = require('pg');
 
 var invitesService = {};
 
@@ -77,82 +76,97 @@ var sendNotification = function (user, req) {
 
 var saveInvite = function (user, stream) {
     return new Promise(function (resolve, reject) {
-        var client = new pg.Client(appVars.postgres.uri);
-        client.connect();
-        client.query("INSERT INTO stream_invites (stream_id, user_id) SELECT $1, $2", [stream.id, user.id],
-            function (err, result) {
-                if (err) {
-                    reject(err);
-                }
-                else {
-                    resolve(null);
-                }
+        var pool = appVars.pool;
+        pool.connect(function (err, client, done) {
+            if(err) {
+                return console.error('error fetching client from pool', err);
             }
-        );
+            client.query("INSERT INTO stream_invites (stream_id, user_id) SELECT $1, $2", [stream.id, user.id],
+                function (err, result) {
+                    done(err);
+                    if (err) {
+                        reject(err);
+                    }
+                    else {
+                        resolve(null);
+                    }
+                }
+            );
+        });
     })
 };
 
 var getInvites = function (user) {
     return new Promise(function (resolve, reject) {
-        var client = new pg.Client(appVars.postgres.uri);
-        client.connect();
-        client.query("SELECT streams.user_id, streams.csync_path, streams.stream_name, streams.description, external_auth.id, external_auth.provider " +
-            "FROM streams INNER JOIN stream_invites ON streams.id = stream_invites.stream_id " +
-            "INNER JOIN external_auth ON streams.user_id = external_auth.user_id " +
-            "WHERE (stream_invites.user_id = $1 AND external_auth.user_id = streams.user_id)", [user.id],
-            function (err, result) {
-                if (err) {
-                    reject(err);
-                }
-                else {
-                    var results = result.rows;
-                    if (results.length == 0) {
-                        resolve(results);
+        var pool = appVars.pool;
+        pool.connect(function (err, client, done) {
+            if (err) {
+                return console.error('error fetching client from pool', err);
+            }
+            client.query("SELECT streams.user_id, streams.csync_path, streams.stream_name, streams.description, external_auth.id, external_auth.provider " +
+                "FROM streams INNER JOIN stream_invites ON streams.id = stream_invites.stream_id " +
+                "INNER JOIN external_auth ON streams.user_id = external_auth.user_id " +
+                "WHERE (stream_invites.user_id = $1 AND external_auth.user_id = streams.user_id)", [user.id],
+                function (err, result) {
+                    done(err);
+                    if (err) {
+                        reject(err);
                     }
                     else {
-                        // flatten external accounts into invites
-                        var invites = {};
+                        var results = result.rows;
+                        if (results.length == 0) {
+                            resolve(results);
+                        }
+                        else {
+                            // flatten external accounts into invites
+                            var invites = {};
 
-                        for (var i = 0; i < results.length; i++) {
-                            var userId = results[i].user_id;
-                            if(invites[userId] == null) {
-                                invites[userId] = results[i];
-                                invites[userId].external_accounts = {};
-                                invites[userId].external_accounts[results[i].provider] = results[i].id;
-                                // clear properties that have been moved to external_accounts
-                                delete invites[userId].id;
-                                delete invites[userId].provider;
+                            for (var i = 0; i < results.length; i++) {
+                                var userId = results[i].user_id;
+                                if (invites[userId] == null) {
+                                    invites[userId] = results[i];
+                                    invites[userId].external_accounts = {};
+                                    invites[userId].external_accounts[results[i].provider] = results[i].id;
+                                    // clear properties that have been moved to external_accounts
+                                    delete invites[userId].id;
+                                    delete invites[userId].provider;
+                                }
+                                else {
+                                    invites[userId].external_accounts[results[i].provider] = results[i].id;
+                                }
                             }
-                            else {
-                                invites[userId].external_accounts[results[i].provider] = results[i].id;
+                            var flattenedInvites = [];
+                            for (var invite in invites) {
+                                flattenedInvites.push(invites[invite])
                             }
+                            resolve(flattenedInvites);
                         }
-                        var flattenedInvites = [];
-                        for (var invite in invites) {
-                            flattenedInvites.push(invites[invite])
-                        }
-                        resolve(flattenedInvites);
                     }
                 }
-            }
-        );
+            );
+        });
     })
 };
 
 var deleteInvites = function (user) {
     return new Promise(function (resolve, reject) {
-        var client = new pg.Client(appVars.postgres.uri);
-        client.connect();
-        client.query("DELETE FROM streams WHERE user_id = $1", [user.id],
-            function (err, result) {
-                if (err) {
-                    reject(err);
-                }
-                else {
-                    resolve();
-                }
+        var pool = appVars.pool;
+        pool.connect(function (err, client, done) {
+            if (err) {
+                return console.error('error fetching client from pool', err);
             }
-        );
+            client.query("DELETE FROM streams WHERE user_id = $1", [user.id],
+                function (err, result) {
+                    done(err);
+                    if (err) {
+                        reject(err);
+                    }
+                    else {
+                        resolve();
+                    }
+                }
+            );
+        });
     })
 };
 

--- a/backend/user/user.controller.js
+++ b/backend/user/user.controller.js
@@ -14,7 +14,6 @@
   limitations under the License.
 */
 
-var pg = require('pg');
 var appVars = require('../config/appVars');
 
 var userController = {};
@@ -25,20 +24,23 @@ userController.registerUser = function(user) {
 
 userController.saveUser = function (user) {
     return new Promise(function (resolve, reject) {
-        var client = new pg.Client(appVars.postgres.uri);
-        client.connect();
-        // Update users if exists, otherwise insert it
-        // Warning: this is not safe if executed from multiple sessions at the same time
-        client.query("UPDATE users SET id=$1, device_token=$2 WHERE id=$1;", [user.id, user.deviceToken]);
-        client.query("INSERT INTO users (id) SELECT $1 WHERE NOT EXISTS (SELECT 1 FROM users WHERE id = $2);",  [user.id, user.id],
-            function (err) {
-                if (err) reject(err);
-
-                resolve(user);
-
-                client.end();
+        var pool = appVars.pool;
+        pool.connect(function (err, client, done) {
+            if (err) {
+                return console.error('error fetching client from pool', err);
             }
-        );
+            // Update users if exists, otherwise insert it
+            // Warning: this is not safe if executed from multiple sessions at the same time
+            client.query("UPDATE users SET id=$1, device_token=$2 WHERE id=$1;", [user.id, user.deviceToken]);
+            client.query("INSERT INTO users (id) SELECT $1 WHERE NOT EXISTS (SELECT 1 FROM users WHERE id = $2);", [user.id, user.id],
+                function (err) {
+                    done(err);
+                    if (err) reject(err);
+
+                    resolve(user);
+                }
+            );
+        });
     });
 };
 
@@ -46,78 +48,97 @@ userController.saveExternalAccount = function(userId, externalAccount) {
     return new Promise(function (resolve, reject) {
         var parameters = [externalAccount.id, externalAccount.accessToken.cipher, externalAccount.accessToken.iv, externalAccount.accessToken.tag,
             externalAccount.refreshToken.cipher, externalAccount.refreshToken.iv, externalAccount.refreshToken.tag, userId, externalAccount.provider];
-        var client = new pg.Client(appVars.postgres.uri);
-        client.connect();
-        // Update external account if exists with that provider, otherwise insert it
-        // Warning: this is not safe if executed from multiple sessions at the same time
-        client.query("UPDATE external_auth SET id=$1, access_token=$2, at_iv=$3, at_tag=$4, refresh_token=$5, rt_iv=$6, rt_tag=$7 WHERE user_id=$8 AND provider=$9;", parameters);
-        client.query("INSERT INTO external_auth (id, access_token, at_iv, at_tag, refresh_token, rt_iv, rt_tag, user_id, provider) SELECT $1, $2, $3, $4, $5, $6, $7, $8, $9"
-            + "WHERE NOT EXISTS (SELECT 1 FROM external_auth WHERE user_id=$8 AND provider=$9);", parameters,
-            function (err) {
-                if (err) reject(err);
-
-                resolve();
-
-                client.end();
+        var pool = appVars.pool;
+        pool.connect(function (err, client, done) {
+            if (err) {
+                return console.error('error fetching client from pool', err);
             }
-        );
+            // Update external account if exists with that provider, otherwise insert it
+            // Warning: this is not safe if executed from multiple sessions at the same time
+            client.query("UPDATE external_auth SET id=$1, access_token=$2, at_iv=$3, at_tag=$4, refresh_token=$5, rt_iv=$6, rt_tag=$7 WHERE user_id=$8 AND provider=$9;", parameters);
+            client.query("INSERT INTO external_auth (id, access_token, at_iv, at_tag, refresh_token, rt_iv, rt_tag, user_id, provider) SELECT $1, $2, $3, $4, $5, $6, $7, $8, $9"
+                + "WHERE NOT EXISTS (SELECT 1 FROM external_auth WHERE user_id=$8 AND provider=$9);", parameters,
+                function (err) {
+                    done(err);
+                    if (err) reject(err);
+
+                    resolve();
+                }
+            );
+        });
     });
 };
 
 userController.getUserByID = function (id) {
   return new Promise(function (resolve, reject) {
-      var client = new pg.Client(appVars.postgres.uri);
-      client.connect();
-      // Get user info
-      client.query("SELECT id, device_token FROM users WHERE id=$1", [id],
-          function (err, result) {
-              if (err) reject(err);
-              if (result.rowCount < 1) {
-                  resolve(null);
-                  client.end();
-              }
-              else {
-                  var user = {id: result.rows[0].id, deviceToken: result.rows[0].device_token};
-                  // Get external accounts
-                  client.query("SELECT id, provider, access_token, at_iv, at_tag FROM external_auth WHERE user_id=$1", [id],
-                      function (err, result) {
-                          if (err) reject(err);
-                          user.externalAccounts = result.rows;
-                          resolve(user);
-
-                          client.end();
-                      });
-              }
+      var pool = appVars.pool;
+      pool.connect(function (err, client, done) {
+          if (err) {
+              return console.error('error fetching client from pool', err);
           }
-      );
+          // Get user info
+          client.query("SELECT id, device_token FROM users WHERE id=$1", [id],
+              function (err, result) {
+                  if (err) reject(err);
+                  if (result.rowCount < 1) {
+                      done(err);
+                      resolve(null);
+                  }
+                  else {
+                      var user = {id: result.rows[0].id, deviceToken: result.rows[0].device_token};
+                      // Get external accounts
+                      client.query("SELECT id, provider, access_token, at_iv, at_tag FROM external_auth WHERE user_id=$1", [id],
+                          function (err, result) {
+                              done(err);
+                              if (err) reject(err);
+                              user.externalAccounts = result.rows;
+                              resolve(user);
+                          });
+                  }
+              }
+          );
+      });
   })
 };
 
 userController.getUserAccountByExternalAccount = function (externalAccount) {
   return new Promise(function (resolve, reject) {
-      var client = new pg.Client(appVars.postgres.uri);
-      client.connect();
-      // Get user info
-      client.query("SELECT user_id FROM external_auth WHERE id=$1 AND provider=$2", [externalAccount.id, externalAccount.provider],
-          function (err, result) {
-              if (err) reject(err);
-              if (result.rowCount < 1) {
-                  resolve(null);
-              }
-              else {
-                  client.query("SELECT id, device_token FROM users WHERE id=$1", [result.rows[0].user_id],
-                      function (err, result) {
-                          if (err) reject(err);
-                          if (result.rowCount < 1) {
-                              resolve(null);
-                              client.end();
-                          }
-                          resolve({id: result.rows[0].id, deviceToken: result.rows[0].device_token});
-                          client.end();
-                      });
-              }
+      var pool = appVars.pool;
+      pool.connect(function (err, client, done) {
+          if (err) {
+              return console.error('error fetching client from pool', err);
           }
-      );
+          // Get user info
+          client.query("SELECT user_id FROM external_auth WHERE id=$1 AND provider=$2", [externalAccount.id, externalAccount.provider],
+              function (err, result) {
+                  if (err) {
+                      done(err);
+                      reject(err);
+                      return
+                  }
+                  if (result.rowCount < 1) {
+                      done(err);
+                      resolve(null);
+                  }
+                  else {
+                      client.query("SELECT id, device_token FROM users WHERE id=$1", [result.rows[0].user_id],
+                          function (err, result) {
+                              done(err);
+                              if (err) {
+                                  done(err);
+                                  reject(err);
+                                  return
+                              }
+                              if (result.rowCount < 1) {
+                                  resolve(null);
+                                  client.end();
+                              }
+                              resolve({id: result.rows[0].id, deviceToken: result.rows[0].device_token});
+                          });
+                  }
+              }
+          );
+      });
   })
 };
 
@@ -182,35 +203,44 @@ userController.getOrCreateStream = function (req) {
     var streamName = req.body["streamName"];
     var streamDescription = req.body["streamDescription"];
     return new Promise(function (resolve, reject) {
-        var client = new pg.Client(appVars.postgres.uri);
-        client.connect();
-        client.query("SELECT * FROM streams WHERE user_id=$1", [user.id],
-            function (err, result) {
-                if (err) {
-                    reject(err);
-                    return;
-                }
-                if (result.rowCount > 0) {
-                    resolve(result.rows[0]);
-                    client.end();
-                }
-                else {
-                    client.query("INSERT INTO streams (user_id, csync_path, stream_name, description) VALUES ($1, $2, $3, $4) RETURNING id", [user.id, streamPath, streamName, streamDescription],
-                        function (err, result) {
-                            if (err) {
-                                reject(err);
-                                return;
-                            }
-                            if (result.rowCount < 1) {
-                                reject(null);
-                                client.end();
-                            }
-                            resolve({id: result.rows[0].id, user_id: user.id, csync_path: streamPath, stream_name: streamName});
-                            client.end();
-                        });
-                }
+        var pool = appVars.pool;
+        pool.connect(function (err, client, done) {
+            if (err) {
+                return console.error('error fetching client from pool', err);
             }
-        );
+            client.query("SELECT * FROM streams WHERE user_id=$1", [user.id],
+                function (err, result) {
+                    if (err) {
+                        done(err);
+                        reject(err);
+                        return;
+                    }
+                    if (result.rowCount > 0) {
+                        done(err);
+                        resolve(result.rows[0]);
+                    }
+                    else {
+                        client.query("INSERT INTO streams (user_id, csync_path, stream_name, description) VALUES ($1, $2, $3, $4) RETURNING id", [user.id, streamPath, streamName, streamDescription],
+                            function (err, result) {
+                                done(err);
+                                if (err) {
+                                    reject(err);
+                                    return;
+                                }
+                                if (result.rowCount < 1) {
+                                    reject(null);
+                                }
+                                resolve({
+                                    id: result.rows[0].id,
+                                    user_id: user.id,
+                                    csync_path: streamPath,
+                                    stream_name: streamName
+                                });
+                            });
+                    }
+                }
+            );
+        });
     })
 };
 
@@ -242,22 +272,25 @@ function generateId() {
 
 function mergeIds(id1, id2) {
     return new Promise(function (resolve, reject) {
-        var client = new pg.Client(appVars.postgres.uri);
-        client.connect();
-        client.query("UPDATE external_auth SET user_id=$1 WHERE user_id = $2;", [id1, id2]);
-        client.query("DELETE FROM users WHERE id=$1;", [id2],
-            function (err) {
-                if (err) reject(err);
-
-                client.end();
-                userController.getUserByID(id1)
-                    .then(function (user) {
-                        resolve(user);
-                    }, function (error) {
-                        reject(error);
-                    });
+        var pool = appVars.pool;
+        pool.connect(function (err, client, done) {
+            if (err) {
+                return console.error('error fetching client from pool', err);
             }
-        );
+            client.query("UPDATE external_auth SET user_id=$1 WHERE user_id = $2;", [id1, id2]);
+            client.query("DELETE FROM users WHERE id=$1;", [id2],
+                function (err) {
+                    done(err);
+                    if (err) reject(err);
+                    userController.getUserByID(id1)
+                        .then(function (user) {
+                            resolve(user);
+                        }, function (error) {
+                            reject(error);
+                        });
+                }
+            );
+        });
     })
 }
 

